fcrepo4/fcrepo4
# Add TimeMap headers and HttpResource converter

**JIRA Ticket**: https://jira.duraspace.org/browse/FCREPO-2623

### ~Don't Merge Yet~
This does not include the RDF body listing of Mementos yet. 
~Also I started by building the HttpResourceConverter changes, they seemed important, but now I'm not sure they aren't just extraneous.~ I removed the HttpResourceConverter stuff, if its needed later it can be added later.

# What does this Pull Request do?
Adds appropriate Memento headers returned for requests to resource/fcr:versions endpoint for the timemap resource.

~Also includes an HttpResourceConverter...but now I'm not sure it is need~

# What's new?
Repurposes the old FedoraVersioning to respond to .../fcr:versions GET/HEAD requests.

# How should this be tested?

But the PR, create a versionable resource (http://localhost:8080/rest/foo) by including a Link header of
`<http://fedora.info/definitions/fcrepo#VersionedResource>; rel="type"`
when you PUT or POST it.

Then GET the resource http://localhost:8080/rest/foo/fcr:versions and ensure that you see the Link headers
`<http://localhost:8080/rest/foo>; rel="original"`
`<http://localhost:8080/rest/foo>; rel="timegate"`

Also:
`Vary-Post: Memento-Datetime`
and
`Accept: POST,GET,HEAD,OPTIONS`

Create a VersionedResource
```
> curl -i -XPUT -H"Link: <http://fedora.info/definitions/fcrepo#VersionedResource>; rel=\"type\"" http://localhost:8080/rest/versionRDF
HTTP/1.1 201 Created
Date: Tue, 24 Oct 2017 14:05:50 GMT
ETag: W/"17d6e9e10b86b405d3474a2b73b26fce8efe461f"
Last-Modified: Tue, 24 Oct 2017 14:05:51 GMT
Link: <http://localhost:8080/static/constraints/ContainerConstraints.rdf>; rel="http://www.w3.org/ns/ldp#constrainedBy"
Link: <http://fedora.info/definitions/fcrepo#VersionedResource>; rel="type"
Link: <http://localhost:8080/rest/versionRDF>; rel="timegate"
Link: <http://localhost:8080/rest/versionRDF/fcr:versions>; rel="timemap"
Location: http://localhost:8080/rest/versionRDF
Content-Type: text/plain
Content-Length: 37
Server: Jetty(9.3.1.v20150714)

http://localhost:8080/rest/versionRDF
```
Get the TimeMap
```
> curl -i -G http://localhost:8080/rest/versionRDF/fcr:versions
HTTP/1.1 200 OK
Date: Tue, 24 Oct 2017 14:06:22 GMT
Link: <http://www.w3.org/ns/ldp#Resource>; rel="type"
Link: <http://www.w3.org/ns/ldp#RDFSource>; rel="type"
Link: <http://fedora.info/definitions/v4/repository#TimeMap>; rel="type"
Vary-Post: Memento-Datetime
Allow: POST,HEAD,GET,OPTIONS
Link: <http://localhost:8080/rest/versionRDF>; rel="original"
Link: <http://localhost:8080/rest/versionRDF>; rel="timegate"
Content-Type: text/turtle;charset=utf-8
Content-Length: 779
Server: Jetty(9.3.1.v20150714)

@prefix premis:  <http://www.loc.gov/premis/rdf/v1#> .
@prefix test:  <info:fedora/test/> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xsi:  <http://www.w3.org/2001/XMLSchema-instance> .
@prefix xmlns:  <http://www.w3.org/2000/xmlns/> .
@prefix rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix fedora:  <http://fedora.info/definitions/v4/repository#> .
```

The RDF body is incorrect, and can be fixed as a follow-on PR once the creation of mementos work is complete.

Example:
* Does this change require documentation to be updated? no
* Does this change add any new dependencies? no
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? no
* Could this change impact execution of existing code?no

# Interested parties
@awoods @birkland @dbernstein @escowles @bseeger
